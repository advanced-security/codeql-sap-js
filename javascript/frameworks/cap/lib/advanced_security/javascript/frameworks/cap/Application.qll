/**
 * Definitions pertaining to the application as a whole.
 */

import javascript
import advanced_security.javascript.frameworks.cap.PackageJson

class RootDirectory extends Folder {
  RootDirectory() {
    exists(PackageJson packageJson | this = packageJson.getJsonFile().getParentContainer())
  }

  /**
   * Gets the path of a file relative to this root directory.
   */
  string getFilePathRelativeToRoot(File file) {
    result = file.getAbsolutePath().regexpReplaceAll(this.getAbsolutePath(), ".") and
    result.charAt(0) = "."
  }

  /**
   * Holds if this root directory of the application contains the given file.
   */
  predicate contains(File file) { exists(this.getFilePathRelativeToRoot(file)) }
}
