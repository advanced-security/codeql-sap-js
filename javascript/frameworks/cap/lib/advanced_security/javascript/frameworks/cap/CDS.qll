import javascript
import DataFlow

module CDS {
  // TODO: should this base type be more specific?
  abstract class ServiceInstance extends DataFlow::Node { }

  /**
   * Call to`cds.serve`
   */
  class CdsServeCall extends ServiceInstance {
    CdsServeCall() { this = any(CdsFacade cds).getMember("serve").getACall() }
  }

  /**
   * call to:
   *  `new cds.ApplicationService` or `new cds.Service`
   */
  class ServiceConstructor extends ServiceInstance {
    ServiceConstructor() { this = any(ApplicationService cds).getAnInstantiation() }
  }

  /**
   * return value of `cds.connect.to`
   */
  class ConnectTo extends ServiceInstance {
    ConnectTo() { this = any(CdsFacade cds).getMember("connect").getMember("to").getACall() }
  }

  /** Last argument to the service methods `srv.before`, `srv.on`, and `srv.after` */
  private class RequestHandler extends FunctionNode { }

  private class ErrorHandler extends RequestHandler { }

  /**
   * Subclassing ApplicationService via `extends`:
   * ```js
   * class SomeService extends cds.ApplicationService
   * ```
   */
  class UserDefinedApplicationService extends ClassNode {
    UserDefinedApplicationService() {
      exists(ApplicationService cdsApplicationService |
        this.getASuperClassNode() = cdsApplicationService.asSource()
      )
    }
  }

  /**
   * Subclassing ApplicationService via `cds.service.impl`:
   * ```js
   * const cds = require('@sap/cds')
   * module.exports = cds.service.impl (function() { ... })
   * ```
   */
  class OldStyleUserDefinedApplicationService extends MethodCallNode {
    OldStyleUserDefinedApplicationService() {
      exists(CdsFacade cds | this = cds.getMember("service").getMember("impl").getACall())
    }
  }

  /**
   *  Parameter of a `srv.with` method call:
   * ```js
   * cds.serve('./srv/cat-service') .with ((srv) => {
   *     srv.on ('READ','Books', (req) => req.reply([...]))
   * })
   * ```
   *
   * TODO expand this to capture request handlers registered inside the function
   */
  class WithCallParameter extends RequestHandler {
    WithCallParameter() {
      exists(MethodCallNode withCall, ServiceInstance svc |
        withCall.getArgument(0) = this and
        withCall.getMethodName() = "with" and
        withCall.getReceiver() = svc
      )
    }
  }

  /**
   * Parameter of request handler phases `_.before`, `_.on` `_.after`:
   * ```js
   * _.on ('READ','Books', (req) => req.reply([...]))
   * ```
   */
  class EventPhaseNodeParam extends ValueNode, ParameterNode {
    MethodCallNode eventPhase;

    EventPhaseNodeParam() {
      exists(FunctionNode handler |
        eventPhase.getMethodName() = ["before", "on", "after"] and
        eventPhase.getLastArgument() = handler and
        handler.getLastParameter() = this
      )
    }

    MethodCallNode getEventPhaseNode() { result = eventPhase }
  }

  /**
   * Parameter of request handler of `srv.on`:
   * ```js
   * this.on ('READ','Books', (req) => req.reply([...]))
   * ```
   * not sure how else to know which service is registering the handler
   */
  class RequestSource extends EventPhaseNodeParam {
    RequestSource() {
      // TODO : consider  - do we need to actually ever know which service the handler is associated to?
      exists(UserDefinedApplicationService svc, FunctionNode init |
        svc.getAnInstanceMember() = init and
        init.getName() = "init" and
        this.getEventPhaseNode().getEnclosingFunction() = init.getAstNode()
      )
      or
      exists(WithCallParameter pa |
        this.getEventPhaseNode().getEnclosingFunction() = pa.getFunction()
      )
    }
  }

  class ApplicationService extends API::Node {
    ApplicationService() { exists(CdsFacade c | this = c.getMember("ApplicationService")) }
  }

  /**
   * ```js
   * const cds = require('@sap/cds')
   * ```
   */
  class CdsFacade extends API::Node {
    CdsFacade() { this = API::moduleImport("@sap/cds") }
  }

  /**
   * Call to`cds.log`
   */
  class CdsLogCall extends API::Node {
    CdsLogCall() { this = any(CdsFacade cds).getMember("log") }
  }

  /**
   * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`
   */
  class CdsLogSink extends DataFlow::Node {
    CdsLogSink() {
      this =
        any(CdsLogCall cdsLog)
            .getACall()
            .getAChainedMethodCall(["trace", "debug", "info", "log", "warn", "error"])
            .getAnArgument()
    }
  }

  /**
   * Methods that parse source strings into a CQL expression
   */
  class ParseSink extends DataFlow::Node {
    ParseSink() {
      this =
        any(CdsFacade cds)
            .getMember("parse")
            .getMember(["expr", "ref", "xpr"])
            .getACall()
            .getAnArgument()
    }
  }
}
