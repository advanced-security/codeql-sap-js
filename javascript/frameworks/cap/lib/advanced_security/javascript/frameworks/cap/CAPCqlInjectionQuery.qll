import javascript
import semmle.javascript.security.dataflow.SqlInjectionCustomizations
import advanced_security.javascript.frameworks.cap.CQL
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.dataflow.FlowSteps

class CqlClauseWithStringConcatParameter instanceof CqlClause {
  CqlClauseWithStringConcatParameter() {
    exists(DataFlow::Node queryParameter |
      (
        if this instanceof CqlInsertClause or this instanceof CqlUpsertClause
        then
          queryParameter = this.getArgument().flow() or
          queryParameter = this.getArgument().flow().(SourceNode).getAPropertyWrite().getRhs()
        else queryParameter = this.getArgument().flow()
      ) and
      exists(StringConcatenation::getAnOperand(queryParameter))
    )
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlShortcutMethodCallWithStringConcat instanceof CqlShortcutMethodCall {
  CqlShortcutMethodCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getAQueryParameter()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlClauseParserCallWithStringConcat instanceof CqlClauseParserCall {
  CqlClauseParserCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getCdlString()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlInjectionConfiguration extends TaintTracking::Configuration {
  CqlInjectionConfiguration() { this = "CQL injection from untrusted data" }

  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }

  override predicate isSink(DataFlow::Node node) {
    exists(CqlRunMethodCall cqlRunMethodCall |
      node = cqlRunMethodCall.(CqlRunMethodCall).getAQueryParameter()
    )
    or
    exists(CqlShortcutMethodCallWithStringConcat queryRunnerCall |
      node = queryRunnerCall.(CqlQueryRunnerCall).getAQueryParameter()
    )
    or
    exists(AwaitExpr await, CqlClauseWithStringConcatParameter cqlClauseWithStringConcat |
      node = await.flow() and
      await.getOperand() = cqlClauseWithStringConcat.(CqlClause).asExpr()
    )
  }

  override predicate isSanitizer(DataFlow::Node node) { node instanceof SqlInjection::Sanitizer }

  override predicate isAdditionalTaintStep(DataFlow::Node start, DataFlow::Node end) {
    exists(CqlClauseParserCallWithStringConcat cqlParseCallWithStringConcat |
      start = cqlParseCallWithStringConcat.(CqlClauseParserCall).getAnArgument() and
      end = cqlParseCallWithStringConcat
    )
    or
    exists(CqlClause cqlClause |
      start = cqlClause.getArgument().flow() and
      end = cqlClause.flow()
    )
  }
}
