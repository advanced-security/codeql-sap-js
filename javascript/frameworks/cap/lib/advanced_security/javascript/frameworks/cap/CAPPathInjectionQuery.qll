/**
 * Exported functions from CAP `cds.utils`.
 * Functions described from:
 * https://www.npmjs.com/package/@sap/cds?activeTab=code
 */

import javascript
import advanced_security.javascript.frameworks.cap.CDSUtils

abstract class UtilsSink extends DataFlow::Node {
  abstract string sinkType();
}

abstract class UtilsAccessedPathSink extends UtilsSink {
  override string sinkType() { result = "unrestricted file read" }
}

abstract class UtilsControlledDataSink extends UtilsSink {
  override string sinkType() { result = "tainted data being written to a file" }
}

abstract class UtilsControlledPathSink extends UtilsSink {
  override string sinkType() { result = "unrestricted file operations" }
}

/**
 * This represents the data in calls as follows:
 * ```javascript
 * await write ({foo:'bar'}) .to ('some','file.json')
 * ```
 * sinks in this example are:
 * ```javascript
 * {foo:'bar'}
 * ```
 */
class WrittenData extends UtilsControlledDataSink {
  WrittenData() { exists(FileWriters fw | fw.getData() = this) }
}

/**
 * This represents the filepath accessed as an input for the data in calls as follows:
 * ```javascript
 * await copy('db/data').to('dist/db/data')
 * ```
 * sinks in this example are:
 * ```javascript
 * 'db/data'
 * ```
 */
class AccessedPath extends UtilsAccessedPathSink {
  AccessedPath() { exists(FileReaderWriters fw | fw.getFromPath() = this) }
}

/**
 * This represents the filepath where data is written or a file operation is performed in calls as follows:
 * ```javascript
 * await write ({foo:'bar'}) .to ('some','file.json')
 * ```
 * sinks in this example are:
 * ```javascript
 * 'some'
 * 'file.json'
 * ```
 */
class ControlledInputPath extends UtilsControlledPathSink {
  ControlledInputPath() {
    exists(FileReaders fw | fw.getPath() = this)
    or
    exists(FileReaderWriters fw | fw.getToPath() = this)
    or
    exists(FileWriters fw | fw.getPath() = this)
    or
    exists(DirectoryWriters dw | dw.getPath() = this)
    or
    exists(DirectoryReaders dr | dr.getPath() = this)
  }
}
