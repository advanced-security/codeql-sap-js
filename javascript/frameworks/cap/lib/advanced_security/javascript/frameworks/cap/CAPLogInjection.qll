import javascript
import DataFlow
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    this = any(CdsFacade cds).getMember("log").getACall() and
    name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof TemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"]
    )
  }
}
