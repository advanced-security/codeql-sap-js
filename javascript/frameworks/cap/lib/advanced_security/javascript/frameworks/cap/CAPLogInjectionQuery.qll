import javascript
import semmle.javascript.dataflow.DataFlow
import semmle.javascript.security.dataflow.LogInjectionQuery
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.CDS
import advanced_security.javascript.frameworks.cap.dataflow.DataFlow

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    exists(CdsFacade cds |
      this = cds.getMember("log").getACall() and
      name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
    )
  }

  string getName() { result = name }
}

class ConstantOnlyTemplateLiteral extends TemplateLiteral {
  ConstantOnlyTemplateLiteral() {
    forall(Expr e | e = this.getAnElement() | e instanceof TemplateElement)
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`.
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"] and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof ConstantOnlyTemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log
    )
  }
}

class CAPLogInjectionConfiguration extends LogInjectionConfiguration {
  CAPLogInjectionConfiguration() { this = "CAP Log Injection" }

  override predicate isSource(DataFlow::Node start) {
    super.isSource(start)
    or
    start instanceof RemoteFlowSource
  }

  override predicate isBarrier(DataFlow::Node node) {
    exists(HandlerParameterData handlerParameterData |
      node = handlerParameterData and
      not handlerParameterData.getType() = ["cds.String", "cds.LargeString"]
    )
  }

  override predicate isSink(DataFlow::Node end) { end instanceof CdsLogSink }
}
