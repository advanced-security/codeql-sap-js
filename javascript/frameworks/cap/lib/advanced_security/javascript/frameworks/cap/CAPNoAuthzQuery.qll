import advanced_security.javascript.frameworks.cap.CDS
import advanced_security.javascript.frameworks.cap.CDL
import advanced_security.javascript.frameworks.cap.Conditionals

abstract class CdlElementWithoutAuthn instanceof CdlElement {
  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

abstract class CdlElementWithoutCdsAuthn extends CdlElementWithoutAuthn instanceof CdlElement {
  CdlElementWithoutCdsAuthn() { this.hasNoCdsAccessControl() }
}

class CdlServiceWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlService { }

class CdlEntityWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlEntity {
  CdlEntityWithoutCdsAuthn() {
    this.belongsToServiceWithNoAuthn()
    or
    exists(CdlEntityWithoutCdsAuthn otherCdlEntityWithoutCdsAuthn |
      this.inherits(otherCdlEntityWithoutCdsAuthn)
    )
  }
}

class CdlActionWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlAction {
  CdlActionWithoutCdsAuthn() { this.belongsToServiceWithNoAuthn() }
}

class CdlFunctionWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlAction {
  CdlFunctionWithoutCdsAuthn() { this.belongsToServiceWithNoAuthn() }
}

class CdlElementProtectionWithHandlerRegistration instanceof HandlerRegistration {
  string toString() { result = super.toString() }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    super.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  CdlElementProtectionWithHandlerRegistration() {
    (
      this.isBefore()
      or
      this.isOn()
    ) and
    exists(Handler handler, ConditionalExprOrStatement exprOrStmt |
      handler = this.getHandler() and
      (
        exprOrStmt = handler.getFunction().getBody() or
        exprOrStmt = handler.getFunction().getABodyStmt()
      )
    |
      exprOrStmt.getConditionExpr().getAChildExpr*().flow() instanceof RequestUserIs and
      exists(CdsRequest req |
        exprOrStmt.getPolarity() = true and
        exprOrStmt.getAnElseBranchExpr() = req.getARejectCall().asExpr()
        or
        exprOrStmt.getPolarity() = false and
        exprOrStmt.getAThenBranchExpr() = req.getARejectCall().asExpr()
      )
    )
  }

  string getEntityName() { result = super.getEntityName() }

  string getAnEventName() { result = super.getAnEventName() }
}

abstract class CdlElementWithJsAuthn instanceof CdlElement {
  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

class CdlServiceWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlService {
  CdlServiceWithJsAuthn() {
    exists(CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = "*"
    )
  }
}

class CdlEntityWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlEntity {
  CdlEntityWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAnEntity() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getEntityName() = this.getUnqualifiedName()
    )
  }
}

class CdlActionWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlAction {
  CdlActionWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAnAction() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = this.getUnqualifiedName()
    )
  }
}

class CdlFunctionWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlFunction {
  CdlFunctionWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAFunction() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = this.getUnqualifiedName()
    )
  }
}

abstract class CdlElementWithoutJsAuthn extends CdlElementWithoutAuthn instanceof CdlElement { }

class CdlServiceWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlService {
  CdlServiceWithoutJsAuthn() { not this instanceof CdlServiceWithJsAuthn }
}

class CdlEntityWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlEntity {
  CdlEntityWithoutJsAuthn() { not this instanceof CdlEntityWithJsAuthn }
}

class CdlActionWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlAction {
  CdlActionWithoutJsAuthn() { not this instanceof CdlActionWithJsAuthn }
}

class CdlFunctionWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlFunction {
  CdlFunctionWithoutJsAuthn() { not this instanceof CdlFunctionWithJsAuthn }
}

/**
 * The access to property `user` of a handler's request.
 */
class RequestUser extends SourceNode instanceof PropRef {
  RequestUser() {
    exists(Handler handler |
      this.getBase().getALocalSource() = handler.getRequest() and
      this.getPropertyName() = "user"
    )
  }
}

class RequestUserIs instanceof MethodCallNode {
  string userRole;

  RequestUserIs() {
    exists(RequestUser requestUser |
      this = requestUser.getAMethodCall("is") and
      userRole = this.getArgument(0).getStringValue()
    )
  }

  string toString() { result = super.toString() }
}
