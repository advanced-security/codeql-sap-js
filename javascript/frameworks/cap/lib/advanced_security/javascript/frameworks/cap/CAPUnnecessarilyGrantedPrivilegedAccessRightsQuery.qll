import javascript
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A reference to an entity that is defined in this application.
 */
class LocalEntityReference instanceof EntityReference {
  LocalEntityReference() { not this instanceof RemoteEntityReference }

  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }

  predicate hasRestrictedAccessControl() {
    exists(RestrictCondition restrict |
      restrict =
        this.(EntityReference).getCqlDefinition().getRestrictAnnotation().getARestrictCondition()
    |
      not restrict.grantsToAnyone(_)
    )
  }
}

/**
 * A reference to an entity that is not defined in this application and
 * read from a service instance that is looked up with the name defined in
 * package.json.
 */
class RemoteEntityReference instanceof EntityReference {
  RemoteEntityReference() { not exists(this.getCqlDefinition()) }

  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

abstract class PrivilegedUserInstance extends DataFlow::Node { }

class CdsUserPrivilegedProperty extends PrivilegedUserInstance instanceof PropRead {
  CdsUserPrivilegedProperty() {
    exists(CdsUser cdsUser, PropRead cdsUserRef |
      cdsUserRef = cdsUser.getInducingNode() and
      cdsUserRef.flowsTo(this.getBase()) and
      this.getPropertyName() = "Privileged"
    )
  }
}

class CustomPrivilegedUser extends ClassNode {
  CustomPrivilegedUser() {
    exists(CdsUser cdsUser | this.getASuperClassNode() = cdsUser.asSource()) and
    exists(FunctionNode init |
      init = this.getInstanceMethod("is") and
      forall(Expr expr | expr = init.asExpr().(Function).getAReturnedExpr() |
        expr.mayHaveBooleanValue(true)
      )
    )
  }
}

class CustomPrivilegedUserInstance extends PrivilegedUserInstance, NewNode {
  CustomPrivilegedUserInstance() {
    exists(CustomPrivilegedUser customPrivilegedUserClass |
      this = customPrivilegedUserClass.getAnInstantiation()
    )
  }
}
