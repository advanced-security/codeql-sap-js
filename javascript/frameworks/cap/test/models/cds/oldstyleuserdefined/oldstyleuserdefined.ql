import javascript
import advanced_security.javascript.frameworks.cap.CDS

from CDS::OldStyleUserDefinedApplicationService svc
select svc