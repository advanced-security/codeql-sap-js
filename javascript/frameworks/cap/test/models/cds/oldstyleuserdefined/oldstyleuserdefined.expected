| oldstyleuserdefined.js:3:18:3:49 | cds.ser ... n(){ }) |
